stdlib/test/builtin/test__hash.mojo

```diff
@@ -0,0 +1,803 @@
+        buckets[h % num_buckets] += 1
+    var unfilled = 0
+    for v in buckets:
+        if v[] == 0:
+            unfilled += 1
+
+    var fill_factor = 1 - unfilled / num_buckets
+    assert_true(
+        fill_factor >= lower_bound,
+        str("Fill factor for {} is {}, provided lower boound was {}").format(
+            label, fill_factor, lower_bound
+        ),
+    )
+
+
+def test_fill_factor():
+    var words = List[String]()
+
+    words = gen_word_pairs[words_ar]()
+    assert_fill_factor["AR"](words, len(words), 0.63)
+    assert_fill_factor["AR"](words, len(words) // 2, 0.86)
+    assert_fill_factor["AR"](words, len(words) // 4, 0.98)
+    assert_fill_factor["AR"](words, len(words) // 13, 1.0)
+
+    assert_fill_factor_old_hash["AR"](words, len(words), 0.62)
+
+    words = gen_word_pairs[words_el]()
+    assert_fill_factor["EL"](words, len(words), 0.63)
+    assert_fill_factor["EL"](words, len(words) // 2, 0.86)
+    assert_fill_factor["EL"](words, len(words) // 4, 0.98)
+    assert_fill_factor["EL"](words, len(words) // 12, 1.0)
+
+    assert_fill_factor_old_hash["EL"](words, len(words), 0.205)
+
+    words = gen_word_pairs[words_en]()
+    assert_fill_factor["EN"](words, len(words), 0.63)
+    assert_fill_factor["EN"](words, len(words) // 2, 0.85)
+    assert_fill_factor["EN"](words, len(words) // 4, 0.98)
+    assert_fill_factor["EN"](words, len(words) // 14, 1.0)
+
+    assert_fill_factor_old_hash["EN"](words, len(words), 0.04)
+
+    words = gen_word_pairs[words_he]()
+    assert_fill_factor["HE"](words, len(words), 0.63)
+    assert_fill_factor["HE"](words, len(words) // 2, 0.86)
+    assert_fill_factor["HE"](words, len(words) // 4, 0.98)
+    assert_fill_factor["HE"](words, len(words) // 14, 1.0)
+
+    assert_fill_factor_old_hash["HE"](words, len(words), 0.2)
+
+    words = gen_word_pairs[words_lv]()
+    assert_fill_factor["LV"](words, len(words), 0.63)
+    assert_fill_factor["LV"](words, len(words) // 2, 0.86)
+    assert_fill_factor["LV"](words, len(words) // 4, 0.98)
+    assert_fill_factor["LV"](words, len(words) // 12, 1.0)
+
+    assert_fill_factor_old_hash["LV"](words, len(words), 0.05)
+
+    words = gen_word_pairs[words_pl]()
+    assert_fill_factor["PL"](words, len(words), 0.63)
+    assert_fill_factor["PL"](words, len(words) // 2, 0.86)
+    assert_fill_factor["PL"](words, len(words) // 4, 0.98)
+    assert_fill_factor["PL"](words, len(words) // 13, 1.0)
+
+    assert_fill_factor_old_hash["PL"](words, len(words), 0.05)
+
+    words = gen_word_pairs[words_ru]()
+    assert_fill_factor["RU"](words, len(words), 0.63)
+    assert_fill_factor["RU"](words, len(words) // 2, 0.86)
+    assert_fill_factor["RU"](words, len(words) // 4, 0.98)
+    assert_fill_factor["RU"](words, len(words) // 13, 1.0)
+
+    assert_fill_factor_old_hash["RU"](words, len(words), 0.18)
+
+
+def main():
+    test_hash_byte_array()
+    test_avalanche()
+    test_trailing_zeros()
+    test_fill_factor()
```

Comment:
That's awesome! Can I assume this is consistent with the Rust tests, so we can trust the algorithm by comparison with Rust?

Reply:
This would be good to investigate, I agree 👍